// react-rainbow-components/components/Textarea/pageObject/index.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Textarea page object class.
 * @class
 */
var PageTextarea = function () {
  /**
   * Create a new Textarea page object.
   * @constructor
   * @param {string} rootElement - The selector of the Textarea root element.
   */
  function PageTextarea(rootElement) {
    _classCallCheck(this, PageTextarea);

    this.rootElement = rootElement;
  }

  /**
   * Clicks the textarea element.
   * @method
   */


  _createClass(PageTextarea, [{
    key: 'click',
    value: function click() {
      $(this.rootElement).$('textarea').click();
    }

    /**
     * Clicks the label element.
     * @method
     */

  }, {
    key: 'clickLabel',
    value: function clickLabel() {
      $(this.rootElement).$('label').click();
    }

    /**
     * Returns true when the textarea element has focus.
     * @method
     * @returns {bool}
     */

  }, {
    key: 'hasFocusTextarea',
    value: function hasFocusTextarea() {
      return $(this.rootElement).$('textarea').isFocused();
    }

    /**
     * It set a value in the textarea.
     * @method
     * @param {string} values - The value to set in the textarea.
     */

  }, {
    key: 'setValue',
    value: function setValue(value) {
      $(this.rootElement).$('textarea').setValue(value);
    }

    /**
     * It get the value of the textarea.
     * @method
     * @returns {string}
     */

  }, {
    key: 'getValue',
    value: function getValue() {
      return $(this.rootElement).$('textarea').getValue();
    }
  }]);

  return PageTextarea;
}();

module.exports = PageTextarea;
