// react-rainbow-components/components/PrimitiveMenu/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.insertChildOrderly = insertChildOrderly;
exports.findItemByKey = findItemByKey;
exports.findItemIndex = findItemIndex;
exports.getChildMenuItemNodes = getChildMenuItemNodes;

function _toArray(arr) { return Array.isArray(arr) ? arr : Array.from(arr); }

function getRefIndex(nodes, ref) {
    return nodes.indexOf(ref);
}

function sortChildren(childrenRefs, nodes) {
    var _childrenRefs = _toArray(childrenRefs),
        newChildrenRefs = _childrenRefs.slice(0);

    newChildrenRefs.sort(function (refA, refB) {
        return getRefIndex(nodes, refA) - getRefIndex(nodes, refB);
    });
    return newChildrenRefs;
}

function insertChildOrderly(childrenRefs, childRef, nodes) {
    var newChildrenRefs = childrenRefs.concat([childRef]);
    var sortedChildren = sortChildren(newChildrenRefs, nodes);
    return sortedChildren;
}

function findItemByKey(key, childrenRefs) {
    return childrenRefs.find(function (child) {
        return child.innerText[0].toLowerCase() === key.toLowerCase();
    });
}

function findItemIndex(childrenRefs, item) {
    return childrenRefs.findIndex(function (child) {
        return child === item;
    });
}

function getChildMenuItemNodes(ref) {
    if (ref) {
        return ref.querySelectorAll('a[role="menuitem"]');
    }
    return [];
}
