// react-rainbow-components/components/Tab/pageObject/index.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Tab page object class.
 * @class
 */
var PageTab = function () {
    /**
     * Create a new Tab page object.
     * @constructor
     * @param {string} rootElement - The selector of the Tab root element.
     */
    function PageTab(rootElement) {
        _classCallCheck(this, PageTab);

        this.rootElement = rootElement;
    }

    /**
     * Clicks the tab item.
     * @method
     */


    _createClass(PageTab, [{
        key: 'click',
        value: function click() {
            $(this.rootElement).$('a').click();
        }

        /**
         * Returns true when the tab item has focus.
         * @method
         * @returns {bool}
         */

    }, {
        key: 'hasFocus',
        value: function hasFocus() {
            return $(this.rootElement).$('a').isFocused();
        }

        /**
         * Returns true when the tab item is selected.
         * @method
         * @returns {bool}
         */

    }, {
        key: 'isSelected',
        value: function isSelected() {
            return $(this.rootElement).$('a').getAttribute('class') === 'rainbow-tab_anchor rainbow-tab--active';
        }

        /**
         * Returns true when the tab item is visible in the viewport.
         * @method
         * @returns {bool}
         */

    }, {
        key: 'isVisibleWithinViewport',
        value: function isVisibleWithinViewport() {
            return $(this.rootElement).$('a').isDisplayedInViewport();
        }
    }]);

    return PageTab;
}();

module.exports = PageTab;
