// react-rainbow-components/components/VerticalItem/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /* eslint-disable no-script-url, react/prop-types */


exports.default = VerticalItem;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _context = require('../VerticalNavigation/context');

var _context2 = require('../VerticalSection/context');

var _context3 = require('../VerticalSectionOverflow/context');

var _RenderIf = require('../RenderIf');

var _RenderIf2 = _interopRequireDefault(_RenderIf);

require('./styles.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Item(props) {
    var label = props.label,
        icon = props.icon,
        name = props.name,
        href = props.href,
        onClick = props.onClick,
        notification = props.notification,
        className = props.className,
        style = props.style,
        selectedItem = props.selectedItem,
        onSelect = props.onSelect,
        entityHeaderId = props.entityHeaderId,
        isExpanded = props.isExpanded;

    var isSelected = name === selectedItem;

    var getContainerClassNames = function getContainerClassNames() {
        return (0, _classnames2.default)('rainbow-vertical-item', {
            'rainbow-vertical-item--active': isSelected
        }, className);
    };

    var getAriaCurrent = function getAriaCurrent() {
        if (isSelected) {
            return 'page';
        }
        return undefined;
    };

    function hanldeOnClick(e) {
        onClick(e);
        onSelect(e, name);
    }

    function resolveTabIndex() {
        if (isExpanded === false) {
            return -1;
        }
        return 0;
    }

    return _react2.default.createElement(
        'li',
        { className: getContainerClassNames(), style: style },
        _react2.default.createElement(
            'a',
            {
                href: href,
                onClick: hanldeOnClick,
                'aria-describedby': entityHeaderId,
                className: 'rainbow-vertical-item_action',
                'aria-current': getAriaCurrent(),
                tabIndex: resolveTabIndex()
            },
            _react2.default.createElement(
                'span',
                { className: 'rainbow-vertical-item_action-label' },
                label
            ),
            _react2.default.createElement(
                _RenderIf2.default,
                { isTrue: !!notification },
                _react2.default.createElement(
                    'span',
                    { className: 'rainbow-vertical-item_notification' },
                    notification
                )
            )
        ),
        _react2.default.createElement(
            _RenderIf2.default,
            { isTrue: !!icon },
            _react2.default.createElement(
                'span',
                { className: 'rainbow-vertical-item_icon' },
                icon
            )
        )
    );
}

/**
 * A text-only link within VerticalNavigationSection or VerticalNavigationOverflow.
 * @category Layout
 */
function VerticalItem(props) {
    return _react2.default.createElement(
        _context.Consumer,
        null,
        function (context) {
            return _react2.default.createElement(
                _context2.Consumer,
                null,
                function (entityHeaderId) {
                    return _react2.default.createElement(
                        _context3.Consumer,
                        null,
                        function (isExpanded) {
                            return _react2.default.createElement(Item, _extends({}, props, context, {
                                entityHeaderId: entityHeaderId,
                                isExpanded: isExpanded
                            }));
                        }
                    );
                }
            );
        }
    );
}

VerticalItem.propTypes = {
    /** The text displayed for the navigation item. */
    label: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),
    /** A unique identifier for the navigation item. */
    name: _propTypes2.default.string.isRequired,
    /** The icon to show if it is passed. It must be a svg icon or a font icon. */
    icon: _propTypes2.default.node,
    /** The URL of the page that the navigation item goes to. */
    href: _propTypes2.default.string,
    /** The notification comoponent to display on the right. */
    notification: _propTypes2.default.node,
    /** Event fired when the item is clicked. */
    onClick: _propTypes2.default.func,
    /** A CSS class for the outer element, in addition to the component's base classes. */
    className: _propTypes2.default.string,
    /** An object with custom style applied for the outer element. */
    style: _propTypes2.default.object
};

VerticalItem.defaultProps = {
    label: '',
    icon: undefined,
    href: 'javascript:void(0);',
    onClick: function onClick() {},
    notification: null,
    className: undefined,
    style: undefined
};
