// react-rainbow-components/components/Button/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _content = require('./content');

var _content2 = _interopRequireDefault(_content);

require('./styles.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Buttons are clickable items used to perform an action.
 */
var Button = function (_Component) {
    _inherits(Button, _Component);

    function Button(props) {
        _classCallCheck(this, Button);

        var _this = _possibleConstructorReturn(this, (Button.__proto__ || Object.getPrototypeOf(Button)).call(this, props));

        _this.buttonRef = _react2.default.createRef();
        return _this;
    }

    _createClass(Button, [{
        key: 'getVariantClassNames',
        value: function getVariantClassNames() {
            var variant = this.props.variant;

            if (variant === 'base') {
                return null;
            }
            return 'rainbow-button--' + variant;
        }
    }, {
        key: 'getClassNames',
        value: function getClassNames() {
            var _props = this.props,
                className = _props.className,
                shaded = _props.shaded,
                variant = _props.variant,
                isLoading = _props.isLoading;

            var isShaded = shaded && (variant === 'neutral' || variant === 'brand' || variant === 'destructive' || variant === 'success');

            return (0, _classnames2.default)('rainbow-button', this.getVariantClassNames(), {
                'rainbow-button--shaded': isShaded,
                'rainbow-button--loading': isLoading && variant !== 'base'
            }, className);
        }
    }, {
        key: 'isDisabled',
        value: function isDisabled() {
            var _props2 = this.props,
                disabled = _props2.disabled,
                isLoading = _props2.isLoading;

            return disabled || isLoading;
        }

        /**
         * Sets focus on the element.
         * @public
         */

    }, {
        key: 'focus',
        value: function focus() {
            this.buttonRef.current.focus();
        }

        /**
         * Sets click on the element.
         * @public
         */

    }, {
        key: 'click',
        value: function click() {
            this.buttonRef.current.click();
        }

        /**
         * Sets blur on the element.
         * @public
         */

    }, {
        key: 'blur',
        value: function blur() {
            this.buttonRef.current.blur();
        }
    }, {
        key: 'render',
        value: function render() {
            var _props3 = this.props,
                style = _props3.style,
                label = _props3.label,
                children = _props3.children,
                tabIndex = _props3.tabIndex,
                onFocus = _props3.onFocus,
                onBlur = _props3.onBlur,
                onClick = _props3.onClick,
                title = _props3.title,
                type = _props3.type,
                ariaHaspopup = _props3.ariaHaspopup,
                id = _props3.id,
                isLoading = _props3.isLoading,
                variant = _props3.variant,
                ariaPressed = _props3.ariaPressed,
                ariaControls = _props3.ariaControls,
                ariaExpanded = _props3.ariaExpanded,
                onKeyDown = _props3.onKeyDown,
                form = _props3.form;


            return _react2.default.createElement(
                'button',
                {
                    'data-id': 'button-element',
                    id: id,
                    className: this.getClassNames(),
                    style: style,
                    disabled: this.isDisabled(),
                    tabIndex: tabIndex,
                    onFocus: onFocus,
                    onBlur: onBlur,
                    onClick: onClick,
                    title: title,
                    type: type,
                    'aria-haspopup': ariaHaspopup,
                    'aria-controls': ariaControls,
                    'aria-expanded': ariaExpanded,
                    'aria-pressed': ariaPressed,
                    onKeyDown: onKeyDown,
                    form: form,
                    ref: this.buttonRef
                },
                _react2.default.createElement(
                    _content2.default,
                    { variant: variant, label: label, isLoading: isLoading },
                    children
                )
            );
        }
    }]);

    return Button;
}(_react.Component);

exports.default = Button;


Button.propTypes = {
    /** The text to be displayed inside the button. */
    label: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),
    /** The content of the button, used to render icon or text elements inside the button.
     * Children take precedence over label. */
    children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.object]),
    /** The variant changes the appearance of the button.
     * Accepted variants include base, neutral, brand, outline-brand, destructive,
     * success, inverse and border-inverse. */
    variant: _propTypes2.default.oneOf(['base', 'neutral', 'brand', 'outline-brand', 'destructive', 'success', 'inverse', 'border-inverse']),
    /** Specifies true when the button has a shadow around it.
     * Only neutral, brand, destructive and success variant can be shaded. */
    shaded: _propTypes2.default.bool,
    /** Displays tooltip text when the mouse moves over the element. */
    title: _propTypes2.default.string,
    /** Specifies the type of button. Valid values are button, reset, and submit. */
    type: _propTypes2.default.oneOf(['button', 'submit', 'reset']),
    /** Specifies whether this button should be displayed in a disabled state.
     * Disabled buttons can't be clicked. */
    disabled: _propTypes2.default.bool,
    /** Specifies the tab order of an element (when the tab button is used for navigating). */
    tabIndex: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]),
    /** The action triggered when the element is clicked. */
    onClick: _propTypes2.default.func,
    /** The action triggered when a keyboard key is pressed. */
    onKeyDown: _propTypes2.default.func,
    /** The action triggered when the element receives focus. */
    onFocus: _propTypes2.default.func,
    /** The action triggered when the element releases focus. */
    onBlur: _propTypes2.default.func,
    /** Indicates that the element has a popup context menu or sub-level menu. */
    ariaHaspopup: _propTypes2.default.bool,
    /** A space-separated list of element IDs that
     * this button controls the contents or presence of. */
    ariaControls: _propTypes2.default.string,
    /** Indicates whether an element the button controls is expanded or collapsed.
     * Valid values are 'true' or 'false'. */
    ariaExpanded: _propTypes2.default.bool,
    /** Indicates that the element has been pressed. */
    ariaPressed: _propTypes2.default.bool,
    /** It must be the id attribute of a form element that the button is associated with. */
    form: _propTypes2.default.string,
    /** A CSS class for the outer element, in addition to the component's base classes. */
    className: _propTypes2.default.string,
    /** An object with custom style applied to the outer element. */
    style: _propTypes2.default.object,
    /** The id of the outer element. */
    id: _propTypes2.default.string,
    /** If it is set to true, then a loading symbol is displayed. */
    isLoading: _propTypes2.default.bool
};

Button.defaultProps = {
    label: undefined,
    children: null,
    variant: 'neutral',
    shaded: false,
    title: undefined,
    type: 'button',
    disabled: false,
    tabIndex: undefined,
    onClick: function onClick() {},
    onKeyDown: function onKeyDown() {},
    onFocus: function onFocus() {},
    onBlur: function onBlur() {},
    ariaHaspopup: undefined,
    className: undefined,
    style: undefined,
    id: undefined,
    isLoading: false,
    ariaPressed: undefined,
    ariaControls: undefined,
    ariaExpanded: undefined,
    form: undefined
};
