// react-rainbow-components/components/RadioGroup/radioItems.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = RadioItems;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _radio = require('./radio');

var _radio2 = _interopRequireDefault(_radio);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function RadioItems(props) {
    var options = props.options,
        ariaDescribedby = props.ariaDescribedby,
        onChange = props.onChange,
        value = props.value,
        name = props.name,
        required = props.required;


    var isChecked = function isChecked(option) {
        return option.value === value;
    };

    return options.map(function (option, index) {
        var key = 'radio-' + index;
        return _react2.default.createElement(_radio2.default, _extends({
            key: key,
            required: required,
            onChange: onChange,
            isChecked: isChecked(option),
            ariaDescribedby: ariaDescribedby,
            name: name
        }, option));
    });
}

RadioItems.propTypes = {
    ariaDescribedby: _propTypes2.default.string,
    value: _propTypes2.default.string,
    options: _propTypes2.default.array.isRequired,
    onChange: _propTypes2.default.func,
    name: _propTypes2.default.string.isRequired,
    required: _propTypes2.default.bool.isRequired
};

RadioItems.defaultProps = {
    ariaDescribedby: undefined,
    value: undefined,
    onChange: function onChange() {}
};
