// react-rainbow-components/components/VerticalSection/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _utils = require('../../libs/utils');

var _context = require('./context');

var _RenderIf = require('../RenderIf');

var _RenderIf2 = _interopRequireDefault(_RenderIf);

require('./styles.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Represents a section within a VerticalNavigation.
 * @category Layout
 */
var VerticalSection = function (_Component) {
    _inherits(VerticalSection, _Component);

    function VerticalSection(props) {
        _classCallCheck(this, VerticalSection);

        var _this = _possibleConstructorReturn(this, (VerticalSection.__proto__ || Object.getPrototypeOf(VerticalSection)).call(this, props));

        _this.entityHeaderId = (0, _utils.uniqueId)('entity-header');
        return _this;
    }

    _createClass(VerticalSection, [{
        key: 'getClassNames',
        value: function getClassNames() {
            var className = this.props.className;

            return (0, _classnames2.default)('rainbow-vertical-section', className);
        }
    }, {
        key: 'render',
        value: function render() {
            var _props = this.props,
                label = _props.label,
                style = _props.style,
                children = _props.children;


            return _react2.default.createElement(
                'div',
                { className: this.getClassNames(), style: style },
                _react2.default.createElement(
                    _RenderIf2.default,
                    { isTrue: !!label },
                    _react2.default.createElement(
                        'h2',
                        { id: this.entityHeaderId, className: 'rainbow-vertical-section_title' },
                        label
                    )
                ),
                _react2.default.createElement(
                    _context.Provider,
                    { value: this.entityHeaderId },
                    _react2.default.createElement(
                        'ul',
                        null,
                        children
                    )
                )
            );
        }
    }]);

    return VerticalSection;
}(_react.Component);

VerticalSection.propTypes = {
    /** The heading text for this section. */
    label: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),
    /** A CSS class for the outer element, in addition to the component's base classes. */
    className: _propTypes2.default.string,
    /** An object with custom style applied for the outer element. */
    style: _propTypes2.default.object,
    /**
     * This prop that should not be visible in the documentation.
     * @ignore
     */
    children: _propTypes2.default.node
};

VerticalSection.defaultProps = {
    label: null,
    className: undefined,
    style: undefined,
    children: null
};

exports.default = VerticalSection;
