// react-rainbow-components/components/Rating/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _ratingItems = require('./ratingItems');

var _ratingItems2 = _interopRequireDefault(_ratingItems);

require('./styles.css');

var _utils = require('../../libs/utils');

var _RenderIf = require('../RenderIf');

var _RenderIf2 = _interopRequireDefault(_RenderIf);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/** @category Form */
var Rating = function (_Component) {
    _inherits(Rating, _Component);

    function Rating(props) {
        _classCallCheck(this, Rating);

        var _this = _possibleConstructorReturn(this, (Rating.__proto__ || Object.getPrototypeOf(Rating)).call(this, props));

        _this.state = {
            value: props.value
        };
        _this.starGroupNameId = (0, _utils.uniqueId)('starGroup');
        _this.handleOnHover = _this.handleOnHover.bind(_this);
        _this.handleOnLeave = _this.handleOnLeave.bind(_this);
        return _this;
    }

    _createClass(Rating, [{
        key: 'getContainerClassNames',
        value: function getContainerClassNames() {
            var className = this.props.className;

            return (0, _classnames2.default)('rainbow-rating_container', className);
        }
    }, {
        key: 'getName',
        value: function getName() {
            var name = this.props.name;

            if (name) {
                return name;
            }
            return this.starGroupNameId;
        }
    }, {
        key: 'handleOnHover',
        value: function handleOnHover(event) {
            if (event.target.value) {
                return this.setState({ value: event.target.value });
            }
            return null;
        }
    }, {
        key: 'handleOnLeave',
        value: function handleOnLeave() {
            return this.setState({ value: this.props.value });
        }
    }, {
        key: 'render',
        value: function render() {
            var _props = this.props,
                style = _props.style,
                onChange = _props.onChange,
                label = _props.label;
            var value = this.state.value;

            return _react2.default.createElement(
                'fieldset',
                {
                    onMouseOver: this.handleOnHover,
                    onMouseLeave: this.handleOnLeave,
                    className: this.getContainerClassNames(),
                    style: style
                },
                _react2.default.createElement(
                    _RenderIf2.default,
                    { isTrue: !!label },
                    _react2.default.createElement(
                        'legend',
                        { className: 'rainbow-rating_label' },
                        label
                    )
                ),
                _react2.default.createElement(_ratingItems2.default, { onChange: onChange, value: value, name: this.getName() })
            );
        }
    }]);

    return Rating;
}(_react.Component);

exports.default = Rating;


Rating.propTypes = {
    /** The value of the rating stars. This value defaults to 0. */
    value: _propTypes2.default.string,
    /** The action triggered when a value attribute changes. */
    onChange: _propTypes2.default.func,
    /** An identifier for the group of radio items. */
    name: _propTypes2.default.string,
    /** The rating label */
    label: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),
    /** A CSS class for the outer element, in addition to the component's base classes. */
    className: _propTypes2.default.string,
    /** An object with custom style applied for the outer element. */
    style: _propTypes2.default.object
};

Rating.defaultProps = {
    value: undefined,
    onChange: function onChange() {},
    name: undefined,
    label: null,
    className: undefined,
    style: undefined
};
