// react-rainbow-components/components/Lookup/pageObject/menuItem.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * LookupMenuItem page object class.
 * @class
 */
var PageLookupMenuItem = function () {
    /**
     * Create a new LookupMenuItem page object.
     * @constructor
     * @param {string} rootElement - The selector of the LookupMenuItem root element.
     */
    function PageLookupMenuItem(rootElement) {
        _classCallCheck(this, PageLookupMenuItem);

        this.rootElement = rootElement;
    }

    /**
     * Clicks the menu item.
     * @method
     */


    _createClass(PageLookupMenuItem, [{
        key: 'click',
        value: function click() {
            $(this.rootElement).click();
        }

        /**
         * It moves the pointer over the menu item.
         * @method
         */

    }, {
        key: 'hover',
        value: function hover() {
            var itemElement = $(this.rootElement + ' > a');
            itemElement.moveTo();
        }

        /**
         * Returns true when the menu item is active.
         * @method
         * @returns {bool}
         */

    }, {
        key: 'isActive',
        value: function isActive() {
            return $(this.rootElement).getAttribute('class') === 'rainbow-lookup_menu-item rainbow-lookup_menu-item--active';
        }

        /**
         * Returns true when the menu item is visible inside the menu container.
         * @method
         * @returns {bool}
         */

    }, {
        key: 'isVisible',
        value: function isVisible() {
            return $(this.rootElement).isDisplayedInViewport();
        }
    }]);

    return PageLookupMenuItem;
}();

module.exports = PageLookupMenuItem;
