// react-rainbow-components/components/Table/head/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _header = require('./header');

var _header2 = _interopRequireDefault(_header);

require('./styles.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Head = function (_PureComponent) {
    _inherits(Head, _PureComponent);

    function Head() {
        _classCallCheck(this, Head);

        return _possibleConstructorReturn(this, (Head.__proto__ || Object.getPrototypeOf(Head)).apply(this, arguments));
    }

    _createClass(Head, [{
        key: 'resolveSortDirection',
        value: function resolveSortDirection(field) {
            var _props = this.props,
                sortDirection = _props.sortDirection,
                defaultSortDirection = _props.defaultSortDirection;

            if (this.isSorted(field)) {
                return sortDirection;
            }
            return defaultSortDirection;
        }
    }, {
        key: 'isSorted',
        value: function isSorted(field) {
            var sortedBy = this.props.sortedBy;

            if (field) {
                return field === sortedBy;
            }
            return false;
        }
    }, {
        key: 'render',
        value: function render() {
            var _this2 = this;

            var _props2 = this.props,
                columns = _props2.columns,
                rest = _objectWithoutProperties(_props2, ['columns']);

            if (columns) {
                return columns.map(function (column, index) {
                    var header = column.header,
                        field = column.field,
                        columnRest = _objectWithoutProperties(column, ['header', 'field']);

                    var key = 'header-' + index;

                    return _react2.default.createElement(_header2.default, _extends({}, rest, columnRest, {
                        key: key,
                        colIndex: index,
                        content: header,
                        sortDirection: _this2.resolveSortDirection(field),
                        isSorted: _this2.isSorted(field),
                        field: field
                    }));
                });
            }
            return null;
        }
    }]);

    return Head;
}(_react.PureComponent);

exports.default = Head;


Head.propTypes = {
    columns: _propTypes2.default.array,
    sortDirection: _propTypes2.default.string,
    defaultSortDirection: _propTypes2.default.string,
    sortedBy: _propTypes2.default.string
};

Head.defaultProps = {
    columns: undefined,
    sortDirection: undefined,
    defaultSortDirection: 'asc',
    sortedBy: undefined
};
