// react-rainbow-components/components/Chart/resolveOptions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = resolveOptions;
function resolveOptions(conditions) {
    var disableAnimations = conditions.disableAnimations,
        disableLines = conditions.disableLines,
        disableCurves = conditions.disableCurves,
        showLegend = conditions.showLegend,
        legendPosition = conditions.legendPosition,
        showStacked = conditions.showStacked,
        maintainAspectRatio = conditions.maintainAspectRatio;


    var options = {
        maintainAspectRatio: maintainAspectRatio,
        legend: {
            display: showLegend,
            position: legendPosition,
            fullWidth: true,
            labels: {
                usePointStyle: true
            }
        }
    };

    if (disableAnimations) {
        options = _extends({}, options, {
            animation: {
                duration: 0
            },
            hover: {
                animationDuration: 0
            },
            responsiveAnimationDuration: 0
        });
    }
    if (disableLines) {
        options = _extends({}, options, {
            showLines: false
        });
    }
    if (disableCurves) {
        options = _extends({}, options, {
            elements: {
                line: {
                    tension: 0
                }
            }
        });
    }
    if (showStacked) {
        options = _extends({}, options, {
            scales: {
                xAxes: [{
                    stacked: true
                }],
                yAxes: [{
                    stacked: true
                }]
            }
        });
    }

    return options;
}
